In [2]:
import pandas as pd

# Data ingestion and analysis

In [3]:
# data ingestion
dataframe= pd.read_csv('Womens Clothing E-Commerce Reviews.csv',index_col=0)
dataframe.head()

Clothing ID  Age                    Title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         Review Text  Rating  Recommended IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive Feedback Count   Division Name Department Name Class Name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

In [4]:
dataframe.shape

(23486, 10)

In [5]:
# data filtering to get data only for Clothing ID 1078 for analysis
df = dataframe[dataframe['Clothing ID'] == 1078]
df.shape



(1024, 10)

In [6]:
df[['Clothing ID','Division Name']]

Clothing ID   Division Name
69            1078  General Petite
90            1078  General Petite
117           1078  General Petite
467           1078         General
470           1078         General
...            ...             ...
22994         1078  General Petite
22998         1078  General Petite
23004         1078         General
23460         1078         General
23465         1078         General

[1024 rows x 2 columns]

In [7]:
df.head()

Clothing ID  Age                   Title  \
69          1078   56    Great summer fabric!   
90          1078   51  Sweet flattering dress   
117         1078   32                     NaN   
467         1078   61    Great sweater dress!   
470         1078   33         Cute, but cheap   

                                           Review Text  Rating  \
69   I really wanted this to work. alas, it had a s...       3   
90   I love cute summer dresses and this one, espec...       4   
117  This is the perfect summer dress. it can be dr...       5   
467  Nice fit and flare style, not clingy at all. i...       5   
470  When i first opened this dress and tried it on...       3   

     Recommended IND  Positive Feedback Count   Division Name Department Name  \
69                 0                        1  General Petite         Dresses   
90                 1                        0  General Petite         Dresses   
117                1                        2  General Petite         Dresses   
467                1                        1         General         Dresses   
470                0                        0         General         Dresses   

    Class Name  
69     Dresses  
90     Dresses  
117    Dresses  
467    Dresses  
470    Dresses

In [8]:
df.count()

Clothing ID                1024
Age                        1024
Title                       871
Review Text                 987
Rating                     1024
Recommended IND            1024
Positive Feedback Count    1024
Division Name              1024
Department Name            1024
Class Name                 1024
dtype: int64

Out of 1024 rows, only 871 have title and 987 have review text. Thus checking if there exists any with both title & review text missing

In [9]:
df_null = df[df['Title'].isnull() & df['Review Text'].isnull()]
print("->Shape of null value dataframe:",df_null.shape)
df_null.head()

->Shape of null value dataframe: (37, 10)


Clothing ID  Age Title Review Text  Rating  Recommended IND  \
554          1078   45   NaN         NaN       5                1   
3976         1078   35   NaN         NaN       5                1   
4308         1078   29   NaN         NaN       5                1   
4370         1078   40   NaN         NaN       5                1   
4441         1078   69   NaN         NaN       5                1   

      Positive Feedback Count Division Name Department Name Class Name  
554                         0       General         Dresses    Dresses  
3976                        0       General         Dresses    Dresses  
4308                        0       General         Dresses    Dresses  
4370                        0       General         Dresses    Dresses  
4441                        0       General         Dresses    Dresses

So there are 37 values where both Title and Review Text are absent, we shall be classifying such as seperately as "No text to claasify"

Done, Now we will be building a new column to df where we merge text feedbacks fields (Title & Review Text)

# Extracting label to create corpus and extract labels 

In [10]:
df = df.fillna("")
df['joint_feedback'] = df['Title'] +' '+ df['Review Text']
df.head()

Clothing ID  Age                   Title  \
69          1078   56    Great summer fabric!   
90          1078   51  Sweet flattering dress   
117         1078   32                           
467         1078   61    Great sweater dress!   
470         1078   33         Cute, but cheap   

                                           Review Text  Rating  \
69   I really wanted this to work. alas, it had a s...       3   
90   I love cute summer dresses and this one, espec...       4   
117  This is the perfect summer dress. it can be dr...       5   
467  Nice fit and flare style, not clingy at all. i...       5   
470  When i first opened this dress and tried it on...       3   

     Recommended IND  Positive Feedback Count   Division Name Department Name  \
69                 0                        1  General Petite         Dresses   
90                 1                        0  General Petite         Dresses   
117                1                        2  General Petite         Dresses   
467                1                        1         General         Dresses   
470                0                        0         General         Dresses   

    Class Name                                     joint_feedback  
69     Dresses  Great summer fabric! I really wanted this to w...  
90     Dresses  Sweet flattering dress I love cute summer dres...  
117    Dresses   This is the perfect summer dress. it can be d...  
467    Dresses  Great sweater dress! Nice fit and flare style,...  
470    Dresses  Cute, but cheap When i first opened this dress...

In [11]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
nltk.download("wordnet")
nltk.download("omw-1.4")

wnl = WordNetLemmatizer()
corpus = []
for i in df.index:
    review = re.sub('[^a-zA-Z]', ' ', df['joint_feedback'][i])
    review = review.lower()
    review = review.split()
    review = [wnl.lemmatize(word, pos="v") for word in review]
    review = ' '.join(review)
    corpus.append(review)

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\prish\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\prish\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\prish\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [12]:
print(df['joint_feedback'][69])
print(corpus[0])

Great summer fabric! I really wanted this to work. alas, it had a strange fit for me. the straps would not stay up, and it had a weird fit under the breast. it worked standing up, but the minute i sat down it fell off my shoulders. the fabric was beautiful! and i loved that it had pockets.
great summer fabric i really want this to work alas it have a strange fit for me the strap would not stay up and it have a weird fit under the breast it work stand up but the minute i sit down it fell off my shoulder the fabric be beautiful and i love that it have pocket


In [13]:
import spacy

labels = dict()
temp =[] 
nlp = spacy.load("en_core_web_sm")
for sentence in corpus :
  doc = nlp(sentence)
  temp.clear()
  for token in doc:
      if str(token.dep_) == 'amod':
               tag= str(token.text)+ " " + str(token.head.text)
               temp.append(tag)
               if tag in labels:
                     labels[tag] += 1
               else:
                     labels[tag] = 1
      

In [14]:
# filter out elements with <= 2 counts
filtered_labels = dict(filter(lambda x: labels[x[0]] >= 3, labels.items() ))
print(filtered_labels)

{'great fabric': 5, 'flatter dress': 4, 'cute dress': 34, 'perfect dress': 32, 'great dress': 71, 'nice style': 3, 'fit style': 3, 'fit perfect': 8, 'other review': 18, 'fun dress': 8, 'easy dress': 13, 'mid thigh': 3, 'regular xs': 8, 'knit dress': 4, 'long sleeves': 4, 'fit flare': 4, 'beautiful color': 8, 'small petite': 3, 'pretty fabric': 4, 'nice dress': 14, 'only problem': 4, 'local store': 10, 'blue dress': 6, 'blue one': 4, 'fit be': 5, 'petite size': 19, 'first time': 14, 'small frame': 3, 'lovely dress': 15, 'beautiful dress': 57, 'nice fabric': 5, 'casual dress': 8, 'tunic dress': 4, 'comfortable dress': 17, 'other color': 8, 'comfortable flatter': 4, 'high quality': 15, 'loose bite': 5, 'more color': 10, 'jean jacket': 6, 'good quality': 7, 'most dress': 3, 'right place': 8, 'maxi dress': 8, 'new dress': 8, 'many compliment': 22, 'great piece': 5, 'next year': 3, 'other reviewers': 30, 'gorgeous detail': 3, 'sheer sleeves': 5, 'soft material': 4, 'same time': 9, 'nice deta

In [15]:
len(filtered_labels)

196

# Creating aand filtering lables

In [16]:
# filter out labels with relation to ["e-commerce","cloth"] < 0.2 and synonyms already present by checking list 
# function declarations -> filter out labels with relation to 'target words' ["e-commerce","dress","product"] <= 0.35 and eliminatinng synonynous pairs
from sentence_transformers import SentenceTransformer, util 
import numpy as np 
from nltk.corpus import wordnet


model = SentenceTransformer('all-MiniLM-L6-v2')
def synonym_text(word):
       synonyms = [word]
       for syn in wordnet.synsets(word):
              for lm in syn.lemmas():
                     synonyms.append(lm.name())
       return list(set(synonyms))

final_label_list_split = []
def similarity_context(label):
       cos_sim = util.cos_sim(model.encode(["ecommerce","product","dress"]), model.encode(label))
       cos_sim = [float(x) for x in cos_sim]
       return max(cos_sim)
       

def synonym_labels(label):      
       cos_sim_2 = [0]
       adj,topic = label[0].split(' ')
       if final_label_list_split != []:
              for x in final_label_list_split:
                     if x[0] in synonym_text(adj) and x[1] in synonym_text(topic):
                            return False
                     
       if  similarity_context(label[0])>0.35:
              final_label_list_split.append((adj,topic))
              return True
       else:
              return False
       

c:\Users\prish\AppData\Local\Programs\Python\Python311\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [17]:
# preparation for filtering out labels based on  relevance 
# sorting labels based on thier similarity, so we encounter most contextual relevant (symentically close to target words) first and eliminate their synonyms 
sorted_labels = sorted(filtered_labels, key= similarity_context,reverse=True)
print(sorted_labels)
for i in range(len(sorted_labels)):
    sorted_labels[i] = (sorted_labels[i],filtered_labels[sorted_labels[i]])
# Filtering out labels based on  relevance
final_label_list_split = []
sim_filtered_labels = dict(filter(synonym_labels, sorted_labels ))
print("Filtered labels post semantic checks: ",list(sim_filtered_labels.keys()))
print("Length of filtered labels after similarity with online dresses and synonyms grouping:",len(sim_filtered_labels))



['comfy dress', 'comfortable dress', 'more dress', 'fun dress', 'fit dress', 'in dress', 'casual dress', 'simple dress', 'other dress', 'pretty dress', 'most dress', 'loose dress', 'new dress', 'best dress', 'good dress', 'perfect dress', 'white dress', 'cute dress', 'unique dress', 'feminine dress', 'pink dress', 'versatile dress', 'tunic dress', 'easy dress', 'blue dress', 'adorable dress', 'flatter dress', 'super dress', 'beautiful dress', 'entire dress', 'little dress', 'favorite dress', 'maxi dress', 'lovely dress', 'nice dress', 'knit dress', 'great dress', 'gorgeous dress', 'awesome dress', 'fit style', 'jean jacket', 'local retailer', 'long torso', 'fit be', 'comfy fit', 'comfortable fabric', 'sheer sleeves', 'pretty fabric', 'small waist', 'loose fit', 'local store', 'great buy', 'larger size', 'real life', 'top part', 'much material', 'full price', 'high quality', 'small size', 'small fit', 'much fabric', 'appropriate work', 'unique style', 'medium fit', 'looser fit', 'bigger

In [18]:
# Creating a dict based final label based final_label_list_split 
# dict key-> topic(head of adjective) , value -> dict with label values, synonyms to topic and encoder to all labels (to be used later)


final_labels = dict()
for adj,topic in final_label_list_split:
    if topic not in final_labels:
        final_labels[topic] = { 'synonyms': synonym_text(topic) ,"labels" : [adj+ " " +topic]}
    else:
        final_labels[topic]['labels'].append(adj+ " " +topic)

In [19]:
for topic in final_labels:
    final_labels[topic]['encoder'] = model.encode(final_labels[topic]['labels'] )

# Trying test run in the data to see if we are getting sensible results.

In [20]:
topic_list = list(final_labels.keys())

result_list=[ [] for x in corpus]


for index in range(len(corpus)):
# using division, dept and class to get little more relevant to topic  
    joint_feedback = corpus[index]

    if joint_feedback != "":
        split_feedback = joint_feedback.split(' ')
        encoded_feedback = model.encode(joint_feedback)
        
        # following code loops through synonyms of all topics present in 'final_labels', 
        # if any topic / synonym is present -> we shall use cosine similarity to find most semantically similar label
        
        for t in topic_list:
            synonyms_t = synonym_text(t)
            for synonym_t in final_labels[t]['synonyms']:
                if synonym_t in split_feedback:
                    
                    # encoder and cofidence testing 
                    winners = []
                    cos_sim = util.cos_sim(encoded_feedback,final_labels[t]['encoder'])
                    for arr in cos_sim:
                        for i, each_val in enumerate(arr):
                            winners.append([final_labels[t]['labels'][i],each_val])

                    label = sorted(winners, key=lambda x: x[1], reverse=True)[0][0]
                    confidence = sorted(winners, key=lambda x: x[1], reverse=True)[0][1]
                    if confidence > 0.4:
                        result_list[index].append(label)
                    break

        if result_list[index] == [] :
            # Edge case handing for no context label found / no context label above confidence threshold   
            result_list[index] = "Context not found"
        else:
            result_list[index] = '|'.join(result_list[index])
    else:
        # Edge case handing for no text feedback given
        result_list[index] = "No text to analyse"
    
    

In [21]:
df['Context Labels'] = result_list
df.to_csv('trial_run_1078.csv',index=False,header=True)